{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "- Identify disease biomarkers.\n",
    "\n",
    "- Build an inductive model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset Selection\n",
    "\n",
    "- Datasets: 191\n",
    "\n",
    "- Samples: 29.198\n",
    "\n",
    "### OpenRefine:\n",
    "A powerful free, open source tool for working with messy data: cleaning it, transforming it from one format into another and extending it with web services and external data.\n",
    "\n",
    "- Retrieve platform *technology* from [Gene Expression Omnibus](https://0-www-ncbi-nlm-nih-gov.brum.beds.ac.uk/geo/browse/?view=platforms):\n",
    "\n",
    "  - high-throughput sequencing (64 datasets: 1473 samples)\n",
    "\n",
    "  - spotted oligonucleotide (35 datasets: 24.119 samples)\n",
    "\n",
    "  - in situ oligonucleotide (61 datasets: 2987 samples)\n",
    "\n",
    "  - RT-PCR (28 datasets: 544 samples)\n",
    "\n",
    "  - other (3 datasets: 75 samples)\n",
    "\n",
    "- Delete datasets:\n",
    "\n",
    "  - *technology* = other | [RT-PCR](https://biosistemika.com/blog/qpcr-microarrays-rna-sequencing-choose-one)\n",
    "\n",
    "  - *dataset* = EMIR00000180 (64 samples) ([samples in common with EMIR00000177 but inconsistent values](./inconsistency.numbers))\n",
    "\n",
    "  - *dataset* = EMIR00000091 | EMIR00000092 | EMIR00000097 | EMIR00000101 (partitioning of EMIR00000102)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata Preprocessing\n",
    "\n",
    "- Features: 267\n",
    "\n",
    "### OpenRefine\n",
    "\n",
    "- Add dataset-level features.\n",
    "\n",
    "- Split and transpose key:value features.\n",
    "\n",
    "- Delete samples:\n",
    "\n",
    "   - *organism* ≠ Homo sapiens (22 samples)\n",
    "\n",
    "   - *channel_count* ≠ 1 (189 samples)\n",
    "\n",
    "   - *molecule* ≠ total RNA (2 samples)\n",
    "\n",
    "   - EMIR00000090 | EMIR00000008 rows whose disease status is not specified (13 samples, 4 samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "def keep_first_occurrence(row):\n",
    "    \"\"\"\n",
    "    Delete repeated occurrences of the same value in a row.\n",
    "    \"\"\"\n",
    "    new_row = []\n",
    "    for value in row:\n",
    "        new_row.append(value if value not in new_row else None)\n",
    "    return pd.Series(new_row, index=row.index)\n",
    "\n",
    "\n",
    "metadata = pd.read_csv('data/PD.csv', dtype=str).apply(keep_first_occurrence, axis=1)\n",
    "\n",
    "# Remove columns containing a unique value for all the rows\n",
    "metadata.drop(metadata.columns[metadata.apply(pd.Series.nunique) == 1], axis=1, inplace=True)\n",
    "\n",
    "# Remove empty columns\n",
    "metadata.dropna(axis=1, how='all', inplace=True)\n",
    "\n",
    "metadata.to_csv('data/PD.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### OpenRefine\n",
    "\n",
    "- Merge equivalent features associated to different datasets.\n",
    "\n",
    "- Rename features consistently.\n",
    "\n",
    "- Delete irrelevant or redundant features (e.g. *state* has an almost 1:1 relationship with *city*).\n",
    "\n",
    "- Delete features where missing values ⪆ 98%.\n",
    "\n",
    "- Refine dataset-level features (based on majority class) using sample-level data:\n",
    "\n",
    "  - *sample_type*\n",
    "  \n",
    "  - *disease*\n",
    "\n",
    "  - *disease_type*\n",
    "\n",
    "  - *body_site*\n",
    "        \n",
    "- Detect *healthy* samples in each dataset exploiting sample description.\n",
    "\n",
    "- Extract *metastasis* state and *tumor_stage* from sample description.\n",
    "\n",
    "- Standardize *age*, *sex*, *city*, *disease*, *disease_type* values."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# miRNA Preprocessing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "def transform_miRNA_name(name):\n",
    "    \"\"\"\n",
    "    Delete whitespaces, standardize separators and remove version suffix from miRNA names.\n",
    "    \"\"\"\n",
    "    name = name.replace(' ', '')\n",
    "    name = name.replace(',hsa', '|hsa')\n",
    "    name = name.replace('+hsa', '|hsa')\n",
    "    name = name.replace('/hsa', '|hsa')\n",
    "    name = name.replace('_hsa', '|hsa')\n",
    "    name = name.split('_v')[0]\n",
    "    return name\n",
    "\n",
    "\n",
    "# Create a dictionary {dataset_name : sample_list}\n",
    "dictionary = pd.read_csv('data/PD.csv').groupby('dataset')['sample_id'].agg(list).to_dict()\n",
    "\n",
    "df_list = []\n",
    "\n",
    "for dataset in dictionary.keys():\n",
    "\n",
    "    df = pd.read_csv(f'data/raw/exp/{dataset}_exp.csv', converters={0: transform_miRNA_name})\n",
    "\n",
    "    # Delete rows indexed by the same miRNA name (dataset anomaly)\n",
    "    df = df.drop_duplicates(df.columns[0])\n",
    "\n",
    "    # Delete rows indexed by miRNA names containing 'plus', 'novel' (unknown miRNA), 'hsa-miR-647+a703' (dataset anomaly), '_var2' (dataset contains identical suffix-free miRNA)\n",
    "    df = df[df[df.columns[0]].str.contains('plus|novel|hsa-miR-647\\+a703|_var2') == False]\n",
    "\n",
    "    # Transpose data\n",
    "    df = df.set_index(df.columns[0]).T\n",
    "\n",
    "    # Remove samples not included in metadata\n",
    "    df = df.loc[dictionary[dataset]]\n",
    "    \n",
    "    df_list.append(df)\n",
    "\n",
    "# Concatenate data into a single dataframe\n",
    "values = pd.concat(df_list)\n",
    "\n",
    "# Save miRNA list and frequency\n",
    "pd.DataFrame(values.count(axis=0), columns=['samples']).to_csv('data/miRNA.csv')\n",
    "\n",
    "values.to_csv('data/EXP.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- [Plus](https://www.tecan.com/hubfs/Tecan_Journal/200702/TecanJournal2007-02_page_16.pdf)\n",
    "\n",
    "- [Novel](https://academic.oup.com/bfg/article/15/1/65/1742523)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Issues\n",
    "\n",
    "\n",
    "- Previous experiments: *disease* attribute extracted from dataset table but each dataset also contains healthy samples.\n",
    "\n",
    "- Dataset integration is not recommended. If values in multiple datasets are not comparable, generated results are biased:\n",
    "\n",
    "  - Different datasets use different numeric range: [0,1], [-∞,0], [0,+∞], [-∞,+∞]\n",
    "  \n",
    "  - Model will map dataset numeric range to disease\n",
    "\n",
    "- Select diseases:\n",
    "\n",
    "  - Relevance\n",
    "\n",
    "  - Frequency threshold\n",
    "\n",
    "- Select miRNA:\n",
    "\n",
    "  - Frequency threshold\n",
    "    \n",
    "  - Multiple: ...|...|...\n",
    "\n",
    "  - Suffix: *\n",
    "\n",
    "- Standardize *tumor_stage* and *metastasis*.\n",
    "\n",
    "- Define graph structure."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
